File: alumnium/drivers/appium_driver.py
from contextlib import contextmanager
from time import sleep

from appium.webdriver import Remote
from appium.webdriver.webelement import WebElement
from appium.webdriver.common.appiumby import AppiumBy as By
from appium.webdriver.extensions.action_helpers import ActionHelpers
from selenium.webdriver.common.action_chains import ActionChains
from selenium.webdriver.common.keys import Keys

from alumnium.accessibility import XCUITestAccessibilityTree
from alumnium.logutils import get_logger
from alumnium.screenshot_utils import get_area_screenshot_from_rect

from .base_driver import BaseDriver
from .keys import Key

logger = get_logger(__name__)


class AppiumDriver(BaseDriver):
    def __init__(self, driver: Remote):
        self.driver = driver
        self.autoswitch_to_webview = True
        self.delay = 0
        self.hide_keyboard_after_typing = False

    @property
    def accessibility_tree(self) -> XCUITestAccessibilityTree:
        sleep(self.delay)
        return XCUITestAccessibilityTree(self.driver.page_source)

    @accessibility_tree.setter
    def accessibility_tree(self, tree: XCUITestAccessibilityTree):
        # This allows us to update the tree when needed
        self._accessibility_tree = tree

    def click(self, id: int):
        self._find_element(id).click()

    def drag_and_drop(self, from_id: int, to_id: int) -> ActionHelpers:
        self.driver.drag_and_drop(self._find_element(from_id), self._find_element(to_id))

    def press_key(self, key: Key):
        keys = []
        if key == Key.BACKSPACE:
            keys.append(Keys.BACKSPACE)
        elif key == Key.ENTER:
            keys.append(Keys.ENTER)
        elif key == Key.ESCAPE:
            keys.append(Keys.ESCAPE)
        elif key == Key.TAB:
            keys.append(Keys.TAB)

        ActionChains(self.driver).send_keys(*keys).perform()

    def quit(self):
        self.driver.quit()

    @property
    def screenshot(self) -> str:
        return self.driver.get_screenshot_as_base64()

    def area_screenshot(self, id: int) -> str:
        """
        Take a screenshot of a specific element by ID.

        For Appium, we use the element's location and size to create a rectangle,
        then crop the full screenshot accordingly.

        Args:
            id: The element ID from the accessibility tree

        Returns:
            Base64 encoded PNG screenshot of the element area
        """
        try:
            element = self._find_element(id)

            try:
                return element.screenshot_as_base64
            except Exception as e:
                logger.debug(f"Native element screenshot failed, using cropping: {e}")
            # Get element's location and size
            rect = {
                'x': element.location['x'],
                'y': element.location['y'],
                'width': element.size['width'],
                'height': element.size['height']
            }

            # Take a full screenshot and crop it
            full_screenshot = self.driver.get_screenshot_as_png()
            logger.info("Taking full screenshot and cropping to element bounds")
            return get_area_screenshot_from_rect(full_screenshot, rect)

        except Exception as e:
            logger.debug(f"Error getting element screenshot for ID {id}: {e}")
            # If element identification fails, refresh the accessibility tree
            logger.debug("Refreshing accessibility tree to get updated element")
            # Clear any cached data
            self.accessibility_tree = XCUITestAccessibilityTree(self.driver.page_source)

            # Try again with fresh tree
            element = self._find_element(id)

            try:
                return element.screenshot_as_base64
            except Exception:
                # Get element's location and size
                rect = {
                    'x': element.location['x'],
                    'y': element.location['y'],
                    'width': element.size['width'],
                    'height': element.size['height']
                }

                # Take a full screenshot and crop it
                full_screenshot = self.driver.get_screenshot_as_png()
                logger.info("Taking full screenshot and cropping to element bounds")
                return get_area_screenshot_from_rect(full_screenshot, rect)

    def select(self, id: int, option: str):
        # TODO: Implement select functionality and the tool
        pass

    def swipe(self, id: int):
        # TODO: Implement swipe functionality and the tool
        pass

    @property
    def title(self) -> str:
        with self.__webview_context() as context:
            if context:
                return self.driver.title
            else:
                return ""

    def type(self, id: int, text: str):
        element = self._find_element(id)
        element.clear()
        element.send_keys(text)
        if self.hide_keyboard_after_typing:
            ActionChains(self.driver).move_to_element(element).move_by_offset(0, -20).click().perform()

    @property
    def url(self) -> str:
        with self.__webview_context() as context:
            if context:
                return self.driver.current_url
            else:
                return ""

    def _find_element(self, id: int) -> WebElement:
        element = self.accessibility_tree.element_by_id(id)
        xpath = f"//{element.type}"

        props = {}
        if element.name:
            props["name"] = element.name
        if element.value:
            props["value"] = element.value
        if element.label:
            props["label"] = element.label

        if props:
            props = [f'@{k}="{v}"' for k, v in props.items()]
            xpath += f"[{' and '.join(props)}]"

        return self.driver.find_element(By.XPATH, xpath)

    @contextmanager
    def __webview_context(self):
        if self.autoswitch_to_webview:
            current_context = self.driver.current_context
            for context in self.driver.contexts:
                if "WEBVIEW" in context:
                    self.driver.switch_to.context(context)
                    yield context
                    self.driver.switch_to.context(current_context)
                    return

        yield None

File: packages/java/src/main/java/ai/alumnium/driver/AppiumDriver.java
package ai.alumnium.driver;

import ai.alumnium.Config;
import ai.alumnium.accessibility.AccessibilityElement;
import ai.alumnium.accessibility.BaseAccessibilityTree;
import ai.alumnium.accessibility.UIAutomator2AccessibilityTree;
import ai.alumnium.accessibility.XCUITestAccessibilityTree;
import ai.alumnium.tool.BaseTool;
import ai.alumnium.tool.ClickTool;
import ai.alumnium.tool.DragAndDropTool;
import ai.alumnium.tool.PressKeyTool;
import ai.alumnium.tool.TypeTool;
import io.appium.java_client.AppiumBy;
import io.appium.java_client.remote.SupportsContextSwitching;
import java.time.Duration;
import java.util.Collections;
import java.util.Set;
import org.openqa.selenium.By;
import org.openqa.selenium.JavascriptExecutor;
import org.openqa.selenium.OutputType;
import org.openqa.selenium.TakesScreenshot;
import org.openqa.selenium.WebElement;
import org.openqa.selenium.interactions.Actions;
import org.openqa.selenium.interactions.Pause;
import org.openqa.selenium.interactions.PointerInput;
import org.openqa.selenium.interactions.PointerInput.Origin;
import org.openqa.selenium.interactions.Sequence;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * Appium (mobile) implementation of {@link BaseDriver}.
 *
 * <p>Switches between the {@code NATIVE_APP} and a web-view context when gathering source vs.
 * querying title/URL. Tools unsupported on mobile (e.g. tab switching, drag slider) throw {@link
 * UnsupportedOperationException}.
 */
public final class AppiumDriver extends BaseDriver {

  private static final Logger LOG = LoggerFactory.getLogger(AppiumDriver.class);

  private static final String WAITER_SCRIPT = loadScript("/ai/alumnium/driver/scripts/waiter.js");
  private static final String WAIT_FOR_SCRIPT =
      loadScript("/ai/alumnium/driver/scripts/waitFor.js");

  public enum Platform {
    UIAUTOMATOR2,
    XCUITEST
  }

  private final io.appium.java_client.AppiumDriver driver;
  private final Platform platform;
  public boolean autoswitchContexts = true;
  public double delay = Config.DELAY;
  public boolean hideKeyboardAfterTyping = false;
  public boolean doubleFetchPageSource = false;
  // Cached positive detection of a web context (WEBVIEW_*/CHROMIUM). Null until first checked.
  private Boolean sessionHasWebContext = null;
  public final Set<Class<? extends BaseTool>> supportedTools =
      Set.of(ClickTool.class, DragAndDropTool.class, PressKeyTool.class, TypeTool.class);

  public AppiumDriver(io.appium.java_client.AppiumDriver driver) {
    this.driver = driver;
    Object automationName = driver.getCapabilities().getCapability("automationName");
    if (automationName != null && automationName.toString().equalsIgnoreCase("uiautomator2")) {
      this.platform = Platform.UIAUTOMATOR2;
    } else {
      this.platform = Platform.XCUITEST;
    }
  }

  @Override
  public String platform() {
    return platform == Platform.UIAUTOMATOR2 ? "uiautomator2" : "xcuitest";
  }

  @Override
  public Set<Class<? extends BaseTool>> supportedTools() {
    return supportedTools;
  }

  @Override
  public BaseAccessibilityTree accessibilityTree() {
    waitForWebPageToLoad();
    ensureNativeContext();
    sleep(delay);
    if (doubleFetchPageSource) {
      driver.getPageSource();
    }
    String xml = driver.getPageSource();
    return platform == Platform.UIAUTOMATOR2
        ? new UIAutomator2AccessibilityTree(xml)
        : new XCUITestAccessibilityTree(xml);
  }

  // region Actions

  @Override
  public void click(int id) {
    ensureNativeContext();
    WebElement element = findElement(id);
    scrollIntoView(element);
    element.click();
  }

  @Override
  public void dragSlider(int id, double value) {
    throw new UnsupportedOperationException("dragSlider not supported for Appium");
  }

  @Override
  public void dragAndDrop(int fromId, int toId) {
    ensureNativeContext();
    WebElement fromElement = findElement(fromId);
    WebElement toElement = findElement(toId);
    scrollIntoView(fromElement);
    touchDragAndDrop(fromElement, toElement);
  }

  @Override
  public void pressKey(Key key) {
    ensureNativeContext();
    CharSequence code =
        switch (key) {
          case BACKSPACE -> org.openqa.selenium.Keys.BACK_SPACE;
          case ENTER -> org.openqa.selenium.Keys.ENTER;
          case ESCAPE -> org.openqa.selenium.Keys.ESCAPE;
          case TAB -> org.openqa.selenium.Keys.TAB;
        };
    new Actions(driver).sendKeys(code).perform();
  }

  @Override
  public void quit() {
    driver.quit();
  }

  @Override
  public void back() {
    driver.navigate().back();
  }

  @Override
  public void visit(String url) {
    driver.get(url);
  }

  @Override
  public String screenshot() {
    return ((TakesScreenshot) driver).getScreenshotAs(OutputType.BASE64);
  }

  @Override
  public void scrollTo(int id) {
    scrollIntoView(findElement(id));
  }

  @Override
  public String title() {
    ensureWebviewContext();
    try {
      return driver.getTitle();
    } catch (RuntimeException e) {
      return "";
    }
  }

  @Override
  public void type(int id, String text) {
    ensureNativeContext();
    WebElement element = findElement(id);
    scrollIntoView(element);
    element.clear();
    element.sendKeys(text);
  }

  @Override
  public String url() {
    ensureWebviewContext();
    try {
      return driver.getCurrentUrl();
    } catch (RuntimeException e) {
      return "";
    }
  }

  @Override
  public String app() {
    var caps = driver.getCapabilities();
    for (String key :
        new String[] {"appPackage", "bundleId", "appium:appPackage", "appium:bundleId"}) {
      Object v = caps.getCapability(key);
      if (v != null && !v.toString().isEmpty()) return v.toString();
    }
    return "unknown";
  }

  @Override
  public WebElement findElement(int id) {
    AccessibilityElement element = accessibilityTree().elementById(id);
    return platform == Platform.XCUITEST ? findElementIos(element) : findElementAndroid(element);
  }

  @Override
  public void executeScript(String script) {
    ensureWebviewContext();
    driver.executeScript(script);
  }

  @Override
  public void switchToNextTab() {
    throw new UnsupportedOperationException("Tab switching not supported for Appium");
  }

  @Override
  public void switchToPreviousTab() {
    throw new UnsupportedOperationException("Tab switching not supported for Appium");
  }

  @Override
  public void printToPdf(String filepath) {
    throw new UnsupportedOperationException("Printing to PDF not supported for Appium");
  }

  // endregion
  // region Internals

  private WebElement findElementIos(AccessibilityElement element) {
    StringBuilder predicate = new StringBuilder();
    predicate.append("type == \"").append(element.type()).append("\"");
    appendPredicate(predicate, "name", element.name());
    appendPredicate(predicate, "value", element.value());
    appendPredicate(predicate, "label", element.label());
    String expr = predicate.toString();
    return driver.findElement(AppiumBy.iOSNsPredicateString(expr));
  }

  private void appendPredicate(StringBuilder out, String key, String value) {
    if (value == null || value.isEmpty()) return;
    out.append(" AND ").append(key).append(" == \"").append(value).append("\"");
  }

  private WebElement findElementAndroid(AccessibilityElement element) {
    StringBuilder xpath = new StringBuilder();
    xpath.append("//").append(element.type());
    StringBuilder preds = new StringBuilder();
    appendXpathPredicate(preds, "resource-id", element.androidResourceId());
    appendXpathPredicate(preds, "bounds", element.androidBounds());
    if (preds.length() > 0) {
      xpath.append('[').append(preds).append(']');
    }
    return driver.findElement(By.xpath(xpath.toString()));
  }

  private void appendXpathPredicate(StringBuilder out, String key, String value) {
    if (value == null || value.isEmpty()) return;
    if (out.length() > 0) out.append(" and ");
    out.append('@').append(key).append("=\"").append(value).append('\"');
  }

  private void ensureNativeContext() {
    if (!autoswitchContexts) return;
    var switcher = (SupportsContextSwitching) driver;
    if (!"NATIVE_APP".equals(switcher.getContext())) {
      switcher.context("NATIVE_APP");
    }
  }

  private void ensureWebviewContext() {
    if (!autoswitchContexts) return;
    var switcher = (io.appium.java_client.remote.SupportsContextSwitching) driver;
    var contexts = new java.util.ArrayList<>(switcher.getContextHandles());
    for (int i = contexts.size() - 1; i >= 0; i--) {
      String ctx = contexts.get(i);
      if (isWebContext(ctx)) {
        switcher.context(ctx);
        return;
      }
    }
  }

  // Hybrid apps expose web contexts as WEBVIEW_<package>; Chrome browser sessions expose a single
  // context named CHROMIUM. Anything other than the native context is a web view we can drive.
  private static boolean isWebContext(String context) {
    return context != null && !"NATIVE_APP".equals(context);
  }

  /**
   * For browser / hybrid sessions, switch to the web context and block until the page has
   * stabilized (load complete, network idle, DOM settled) using the shared waiter script, before
   * the native accessibility tree is captured. Native-only sessions are skipped, and timeouts fall
   * through gracefully so a tree is always returned.
   */
  private void waitForWebPageToLoad() {
    if (!autoswitchContexts || !hasWebContext()) return;
    ensureWebviewContext();
    try {
      runWaiter();
    } catch (RuntimeException first) {
      // Retry once, mirroring the Selenium/Playwright drivers.
      try {
        runWaiter();
      } catch (RuntimeException retry) {
        LOG.debug("Failed to wait for web page to stabilize after retry", retry);
      }
    }
  }

  private void runWaiter() {
    JavascriptExecutor js = (JavascriptExecutor) driver;
    js.executeScript(WAITER_SCRIPT);
    Object error = js.executeAsyncScript(WAIT_FOR_SCRIPT);
    if (error != null) {
      LOG.debug("Web page did not stabilize: {}", error);
    }
  }

  private boolean hasWebContext() {
    // Cache positive detection only: a web context can appear after the first capture (hybrid
    // apps), but never disappears once present.
    if (sessionHasWebContext == null || !sessionHasWebContext) {
      var switcher = (SupportsContextSwitching) driver;
      sessionHasWebContext =
          switcher.getContextHandles().stream().anyMatch(AppiumDriver::isWebContext);
    }
    return sessionHasWebContext;
  }

  private void scrollIntoView(WebElement element) {
    if (platform == Platform.UIAUTOMATOR2) {
      scrollIntoViewAndroid(element);
    } else {
      driver.executeScript(
          "mobile: scrollToElement",
          java.util.Map.of(
              "elementId", ((org.openqa.selenium.remote.RemoteWebElement) element).getId()));
    }
  }

  private void scrollIntoViewAndroid(WebElement element) {
    if (element.isDisplayed()) return;
    // Minimal fallback: skip swipe gesture implementation. Real impl
    // should port the Python _scroll_into_view_android loop.
    LOG.warn("Android scroll-into-view fallback is not implemented; element may be off-screen");
  }

  /**
   * Slow viewport-coordinate touch drag for HTML5 drag-and-drop in mobile browsers. Fast
   * element-relative {@link Actions} gestures do not fire drag events.
   */
  private void touchDragAndDrop(WebElement fromElement, WebElement toElement) {
    PointerInput finger = new PointerInput(PointerInput.Kind.TOUCH, "main_pointer");
    Sequence sequence = new Sequence(finger, 0);
    sequence.addAction(
        finger.createPointerMove(Duration.ofMillis(2000), Origin.fromElement(fromElement), 0, 0));
    sequence.addAction(finger.createPointerDown(PointerInput.MouseButton.LEFT.asArg()));
    sequence.addAction(new Pause(finger, Duration.ofMillis(1500)));
    sequence.addAction(
        finger.createPointerMove(Duration.ofMillis(2000), Origin.fromElement(toElement), 0, 0));
    sequence.addAction(new Pause(finger, Duration.ofMillis(1500)));
    sequence.addAction(finger.createPointerUp(PointerInput.MouseButton.LEFT.asArg()));
    driver.perform(Collections.singletonList(sequence));
  }

  private static void sleep(double seconds) {
    if (seconds <= 0) return;
    try {
      Thread.sleep((long) (seconds * 1000d));
    } catch (InterruptedException e) {
      Thread.currentThread().interrupt();
    }
  }

  // endregion
}
